xscen/ensembles.py

```diff
@@ -69,38 +79,158 @@ def ensemble_stats(
+
+        ens_stats = clean_up(
+            ds=ens_stats,
+            common_attrs_only=datasets,
+            common_attrs_open_kwargs=create_kwargs,
         )
-        ens_stats.attrs["cat:id"] = generate_id(df)[0]
 
     ens_stats.attrs["cat:processing_level"] = to_level
 
     return ens_stats
+
+
+def generate_weights(
+    datasets: Union[dict, list],
+    independence_level: str = "all",
+) -> xr.DataArray:
+    """
+    Uses realization attributes to automatically generate weights along the 'realization' dimension.
+
+    Parameters
+    ----------
+    datasets: dict
+        List of Dataset objects that will be included in the ensemble.
+        The datasets should include attributes to help recognize them - 'cat:activity','cat:source', and 'cat:driving_model' for regional models.
+        A dictionary can be passed instead of a list, in which case the keys are used for the 'realization' coordinate.
+        Tip: With a project catalog, you can do: `datasets = pcat.search(**search_dict).to_dataset_dict()`.
+    independence_level: str
+        'all': Weights using the method '1 model - 1 Vote', where every unique combination of 'source' and 'driving_model' is considered a model.
+        'GCM': Weights using the method '1 GCM - 1 Vote'
+
+    Returns
+    -------
+    xr.DataArray
+        Weights along the 'realization' dimension.
+    """
+
+    # TODO: 2-D weights along the horizon dimension
+
+    if (isinstance(datasets, list)) and (isinstance(datasets[0], (str, Path))):
+        raise ValueError(
+            "explicit weights are required if the dataset is a list of paths"
+        )
+
+    # Use metadata to identify the simulation attributes
+    info = {}
+    keys = datasets.keys() if isinstance(datasets, dict) else range(len(datasets))
+    for key in keys:
+        info[key] = {
+            attr.replace("cat:", ""): datasets[key].attrs[attr]
+            for attr in datasets[key].attrs
+            if "cat:" in attr
+        }
+
+    # Prepare an array of 0s, with size == nb. realization
+    weights = xr.DataArray(
+        [0] * len(keys), coords={"realization": ("realization", list(keys))}
+    )
+
+    for r in weights.realization.values:
+
+        # Weight == 0 means it hasn't been processed yet
+        if weights.sel(realization=r) == 0:
+            sim = info[r]
+
+            # Exact group corresponding to the current simulation
+            group = [
+                k
+                for k in info.keys()
+                if (info[k].get("source", None) == sim.get("source", None))
+                and (
+                    info[k].get("driving_model", None) == sim.get("driving_model", None)
+                )
+                and (info[k].get("activity", None) == sim.get("activity", None))
+            ]
+
+            if independence_level == "GCM":
+                if ("driving_model" in sim.keys()) and not (
+                    str(sim.get("driving_model", None)) in (["nan", "None"])
```

Comment:
```suggestion
                    pd.isna(sim.get("driving_model", None))
```
Catches both None and nan.